Statistics

hhio618 opened the pull request:
### Fixes #6 
- Fixed "percentage of humans/bots from the entire pool"
- Countries of users
### Demo
![output](https://user-images.githubusercontent.com/1272158/97274320-bf886b80-1849-11eb-881b-c012b7e1d659.gif)

gaieges reviewed (commented):
Looks pretty good but needs a test to validate that the percentages add up correctly and that we're tracking the correct users (as I see you have under WIP)

hhio618 commented:
Thanks, I'll add up some tests then

uivlis commented:
@hhio618 Any news on the tests?

hhio618 commented:
@uivlis Sorry I was a little busy dealing with covid-19 from last week. I'll try to add tests within hours

hhio618 commented:
Ok, I added some tests, and all tests have been passed. If everything ok I'll upgrade the bot instance on Heroku for more tests.

gaieges commented:
I got a staging deployment working, so we can use that to test to save you some time.  Just make a PR against the `staging` branch and we can test there

gaieges commented:
Lets rebase everything from staging into your branch here.  

Theres are 2 problems with the stats command it looks like, can we fix and get a test to catch these:

first:
```
2020-10-22T18:58:55.185596+00:00 app[web.1]: 2020-10-22 18:58:55,184 - telegram.ext.dispatcher - ERROR - No error handlers are registered, logging exception.
2020-10-22T18:58:55.185599+00:00 app[web.1]: Traceback (most recent call last):
2020-10-22T18:58:55.185599+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1277, in _execute_context
2020-10-22T18:58:55.185600+00:00 app[web.1]: cursor, statement, parameters, context
2020-10-22T18:58:55.185601+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
2020-10-22T18:58:55.185601+00:00 app[web.1]: cursor.execute(statement, parameters)
2020-10-22T18:58:55.185601+00:00 app[web.1]: psycopg2.errors.InFailedSqlTransaction: current transaction is aborted, commands ignored until end of transaction block
2020-10-22T18:58:55.185602+00:00 app[web.1]:
2020-10-22T18:58:55.185602+00:00 app[web.1]:
2020-10-22T18:58:55.185603+00:00 app[web.1]: The above exception was the direct cause of the following exception:
2020-10-22T18:58:55.185603+00:00 app[web.1]:
2020-10-22T18:58:55.185604+00:00 app[web.1]: Traceback (most recent call last):
2020-10-22T18:58:55.185604+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/telegram/ext/dispatcher.py", line 425, in process_update
2020-10-22T18:58:55.185605+00:00 app[web.1]: handler.handle_update(update, self, check, context)
2020-10-22T18:58:55.185605+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/telegram/ext/handler.py", line 145, in handle_update
2020-10-22T18:58:55.185606+00:00 app[web.1]: return self.callback(update, context)
2020-10-22T18:58:55.185606+00:00 app[web.1]: File "/app/app/lib/handlers/base.py", line 16, in func_wrapper
2020-10-22T18:58:55.185607+00:00 app[web.1]: return func(self, *args, **kwargs)
2020-10-22T18:58:55.185607+00:00 app[web.1]: File "/app/app/lib/handlers/stats.py", line 30, in handler
2020-10-22T18:58:55.185608+00:00 app[web.1]: .filter(Channel.chat_id == str(update.message.chat_id))
2020-10-22T18:58:55.185608+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3803, in count
2020-10-22T18:58:55.185608+00:00 app[web.1]: return self.from_self(col).scalar()
2020-10-22T18:58:55.185609+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3523, in scalar
2020-10-22T18:58:55.185609+00:00 app[web.1]: ret = self.one()
2020-10-22T18:58:55.185638+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3490, in one
2020-10-22T18:58:55.185638+00:00 app[web.1]: ret = self.one_or_none()
2020-10-22T18:58:55.185639+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3459, in one_or_none
2020-10-22T18:58:55.185639+00:00 app[web.1]: ret = list(self)
2020-10-22T18:58:55.185640+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3535, in __iter__
2020-10-22T18:58:55.185640+00:00 app[web.1]: return self._execute_and_instances(context)
2020-10-22T18:58:55.185641+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3560, in _execute_and_instances
2020-10-22T18:58:55.185641+00:00 app[web.1]: result = conn.execute(querycontext.statement, self._params)
2020-10-22T18:58:55.185642+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1011, in execute
2020-10-22T18:58:55.185642+00:00 app[web.1]: return meth(self, multiparams, params)
2020-10-22T18:58:55.185642+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 298, in _execute_on_connection
2020-10-22T18:58:55.185643+00:00 app[web.1]: return connection._execute_clauseelement(self, multiparams, params)
2020-10-22T18:58:55.185643+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1130, in _execute_clauseelement
2020-10-22T18:58:55.185644+00:00 app[web.1]: distilled_params,
2020-10-22T18:58:55.185644+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1317, in _execute_context
2020-10-22T18:58:55.185645+00:00 app[web.1]: e, statement, parameters, cursor, context
2020-10-22T18:58:55.185645+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1511, in _handle_dbapi_exception
2020-10-22T18:58:55.185646+00:00 app[web.1]: sqlalchemy_exception, with_traceback=exc_info[2], from_=e
2020-10-22T18:58:55.185646+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
2020-10-22T18:58:55.185646+00:00 app[web.1]: raise exception
2020-10-22T18:58:55.185647+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1277, in _execute_context
2020-10-22T18:58:55.185647+00:00 app[web.1]: cursor, statement, parameters, context
2020-10-22T18:58:55.185647+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
2020-10-22T18:58:55.185648+00:00 app[web.1]: cursor.execute(statement, parameters)
2020-10-22T18:58:55.185649+00:00 app[web.1]: sqlalchemy.exc.InternalError: (psycopg2.errors.InFailedSqlTransaction) current transaction is aborted, commands ignored until end of transaction block
2020-10-22T18:58:55.185649+00:00 app[web.1]:
2020-10-22T18:58:55.185649+00:00 app[web.1]: [SQL: SELECT count(*) AS count_1
2020-10-22T18:58:55.185650+00:00 app[web.1]: FROM (SELECT bots_channels.bot_id AS bots_channels_bot_id, bots_channels.channel_id AS bots_channels_channel_id
2020-10-22T18:58:55.185650+00:00 app[web.1]: FROM bots_channels JOIN channels ON channels.id = bots_channels.channel_id
2020-10-22T18:58:55.185651+00:00 app[web.1]: WHERE channels.chat_id = %(chat_id_1)s) AS anon_1]
2020-10-22T18:58:55.185651+00:00 app[web.1]: [parameters: {'chat_id_1': '1108918469'}]
2020-10-22T18:58:55.185651+00:00 app[web.1]: (Background on this error at: http://sqlalche.me/e/13/2j85)
2020-10-22T18:58:55.183700+00:00 heroku[router]: at=info method=POST path="/1005594849:AAHSLdRmx6XfDB5KCo0PrFQNYh4LEUGd62g" host=staging-hcaptcha-telegram-bot.herokuapp.com request_id=da52bc45-2532-4c6f-87e1-349a8beeb541 fwd="91.108.6.142" dyno=web.1 connect=5ms service=16ms status=204 bytes=148 protocol=https
2020-10-22T18:58:58.558194+00:00 heroku[router]: at=info method=POST path="/1005594849:AAHSLdRmx6XfDB5KCo0PrFQNYh4LEUGd62g" host=staging-hcaptcha-telegram-bot.herokuapp.com request_id=f8dadc81-beb8-4a1c-9047-d12e7b9a458f fwd="91.108.6.142" dyno=web.1 connect=1ms service=3ms status=204 bytes=148 protocol=https
2020-10-22T18:58:58.552961+00:00 app[web.1]: 2020-10-22 18:58:58,552 - telegram.ext.dispatcher - DEBUG - Processing Update: {'update_id': 260546784, 'message': {'message_id': 217, 'date': 1603393138, 'chat': {'id': 1108918469, 'type': 'private', 'username': 'gaieges', 'first_name': 'Evin'}, 'text': '/stats', 'entities': [{'type': 'bot_command', 'offset': 0, 'length': 6}], 'caption_entities': [], 'photo': [], 'new_chat_members': [], 'new_chat_photo': [], 'delete_chat_photo': False, 'group_chat_created': False, 'supergroup_chat_created': False, 'channel_chat_created': False, 'from': {'id': 1108918469, 'first_name': 'Evin', 'is_bot': False, 'username': 'gaieges', 'language_code': 'en'}}}
2020-10-22T18:58:58.553222+00:00 app[web.1]: 10.9.158.174 - - [22/Oct/2020:18:58:58 +0000] "POST /1005594849:AAHSLdRmx6XfDB5KCo0PrFQNYh4LEUGd62g HTTP/1.1" 204 0 "-" "-"
```
second:
```
2020-10-22T18:58:58.562762+00:00 app[web.1]: 2020-10-22 18:58:58,562 - telegram.ext.dispatcher - ERROR - No error handlers are registered, logging exception.
2020-10-22T18:58:58.562764+00:00 app[web.1]: Traceback (most recent call last):
2020-10-22T18:58:58.562764+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/telegram/ext/dispatcher.py", line 425, in process_update
2020-10-22T18:58:58.562765+00:00 app[web.1]: handler.handle_update(update, self, check, context)
2020-10-22T18:58:58.562766+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/telegram/ext/handler.py", line 145, in handle_update
2020-10-22T18:58:58.562766+00:00 app[web.1]: return self.callback(update, context)
2020-10-22T18:58:58.562767+00:00 app[web.1]: File "/app/app/lib/handlers/base.py", line 16, in func_wrapper
2020-10-22T18:58:58.562767+00:00 app[web.1]: return func(self, *args, **kwargs)
2020-10-22T18:58:58.562768+00:00 app[web.1]: File "/app/app/lib/handlers/stats.py", line 40, in handler
2020-10-22T18:58:58.562768+00:00 app[web.1]: percent_humans = num_humans / float(num_all_user) * 100
2020-10-22T18:58:58.562768+00:00 app[web.1]: ZeroDivisionError: float division by zero
```

hhio618 commented:
Good to know. I'll fix these and add another commit

hhio618 commented:
@gaieges I fix and a test case for the second problem but I couldn't reproduce the first problem. Did you get that on a0cc344 commit too?

gaieges commented:
> @gaieges I fix and a test case for the second problem, but I couldn't reproduce the first problem. Did you get that on [a0cc344](https://github.com/hCaptcha/telegram-bot/commit/a0cc3447b49ae2c6c9e57cf7fb544e14de037c01) commit too?

I believe so, yeah; whatever is in the staging branch is what I saw the exception on.

I can try redeploying this version and see if we have any issues like before; could be an open session or something that didn't get closed properly

gaieges commented:
Looks like you have a lint failure here too btw

hhio618 commented:
When I run lint locally, it exits without any error! so I wasn't able to find the lint problem to fix! (https://github.com/hCaptcha/telegram-bot/pull/10/commits/947786b708f44aa1df4a85fb1ab413914154463c).

gaieges commented:
Sure you're using the right version of lint in this repo?  Be sure to have `pipenv install --dev` set up and run it within the pipenv `pipenv run ./bin/lint` .. (by default it should fix it so you should just need to push up changes if its all clean)

gaieges commented:
How goes this @hhio618 ?  Can we close this out soon?

Looks like you fixed the lint but the tests are missing an env var in the test itself (this should be set in the github action env settings, I'm not entirely sure why its not working in this case).  Once we get that fixed we can merge to staging and test, and close out!

hhio618 commented:
@gaieges Just to say that all tests have been passed on my local environment, So from my perspective, it’s ready for merge.

gaieges commented:
> @gaieges Just to say that all tests have been passed on my local environment, So from my perspective, it’s ready for merge.

Do you have the env settings in your github action test config? If so maybe there's a unique aspect of the PR that's not allowing you to set that.

Let's see if running it on the staging branch works; wanna PR there and I'll merge?

hhio618 commented:
I couldn't find env settings in my github action test config @gaieges.

gaieges commented:
https://github.com/hCaptcha/telegram-bot/blob/master/.github/workflows/pytest.yml#L42-L46 <- this just needs to be on your branch.

But if we make a pr against staging I know that it works there, so If tests pass locally it should work there too (and it'll get pushed to staging env)

hhio618 commented:
> https://github.com/hCaptcha/telegram-bot/blob/master/.github/workflows/pytest.yml#L42-L46 <- this just needs to be on your branch.

Do I need to hard-code these or can it be read from the Github secrets?

gaieges commented:
> Do I need to hard-code these or can it be read from the Github secrets?

As long as they're in your branch it should be populating there (unless there's some behavior I'm not aware of there)

gaieges commented:
No luck, I discovered that we need to mock out the tests to not use secrets in that env var.  So I just pushed your content to staging but it failed: https://github.com/hCaptcha/telegram-bot/runs/1328473359

gaieges commented:
Nice that fixed the test but looks like when running `/stats` it fails with this stil (I presume some 0 data in the db):

```
2020-10-29T19:47:21.669847+00:00 app[web.1]: 2020-10-29 19:47:21,668 - telegram.ext.dispatcher - ERROR - No error handlers are registered, logging exception.
2020-10-29T19:47:21.669849+00:00 app[web.1]: Traceback (most recent call last):
2020-10-29T19:47:21.669850+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/telegram/ext/dispatcher.py", line 425, in process_update
2020-10-29T19:47:21.669851+00:00 app[web.1]: handler.handle_update(update, self, check, context)
2020-10-29T19:47:21.669851+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/telegram/ext/handler.py", line 145, in handle_update
2020-10-29T19:47:21.669852+00:00 app[web.1]: return self.callback(update, context)
2020-10-29T19:47:21.669852+00:00 app[web.1]: File "/app/app/lib/handlers/base.py", line 16, in func_wrapper
2020-10-29T19:47:21.669853+00:00 app[web.1]: return func(self, *args, **kwargs)
2020-10-29T19:47:21.669853+00:00 app[web.1]: File "/app/app/lib/handlers/stats.py", line 40, in handler
2020-10-29T19:47:21.669854+00:00 app[web.1]: percent_humans = num_humans / float(num_all_user) * 100
2020-10-29T19:47:21.669854+00:00 app[web.1]: ZeroDivisionError: float division by zero
```

hhio618 commented:
> I discovered that we need to mock out the tests to not use secrets in that env var.   

I think this is a good idea. to achieve that we need to make some changes to whole tests in the repo. maybe I can submit a PR for fixing that in my spare time next week.

> Nice that fixed the test but looks like when running /stats it fails with this stil (I presume some 0 data in the db):   

Perfect. But I can see that bot code is a bit outdated (from the line of error), I fixed this bug and wrote a test case for it.

gaieges commented:
You're right! got the right one deployed, looks like the auto db migration was breaking the deploy. that yellow bot now is updated with the stats help command.  

Getting this now:

```
2020-10-29T22:44:59.750775+00:00 app[web.1]: 2020-10-29 22:44:59,738 - telegram.ext.dispatcher - ERROR - No error handlers are registered, logging exception.
2020-10-29T22:44:59.750792+00:00 app[web.1]: Traceback (most recent call last):
2020-10-29T22:44:59.750793+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/telegram/ext/dispatcher.py", line 425, in process_update
2020-10-29T22:44:59.750793+00:00 app[web.1]: handler.handle_update(update, self, check, context)
2020-10-29T22:44:59.750794+00:00 app[web.1]: File "/app/.heroku/python/lib/python3.7/site-packages/telegram/ext/handler.py", line 145, in handle_update
2020-10-29T22:44:59.750794+00:00 app[web.1]: return self.callback(update, context)
2020-10-29T22:44:59.750795+00:00 app[web.1]: File "/app/app/lib/handlers/base.py", line 16, in func_wrapper
2020-10-29T22:44:59.750795+00:00 app[web.1]: return func(self, *args, **kwargs)
2020-10-29T22:44:59.750796+00:00 app[web.1]: File "/app/app/lib/handlers/stats.py", line 32, in handler
2020-10-29T22:44:59.750796+00:00 app[web.1]: .filter_by(channel_id=channel.id)
2020-10-29T22:44:59.750796+00:00 app[web.1]: AttributeError: 'NoneType' object has no attribute 'id'
```

hhio618 commented:
@gaieges I fixed that and updated test_handle_stats.py file. now it's working as expected!
As Telegram has no get owner id method (for privacy reason maybe), I needed to set the bot owner username (```TELEGRAM_OWNER_USERNAME```) in both configs and .github/workflows/pytest.yml .

gaieges commented on .github/workflows/pytest.yml:
```diff
@@ -45,9 +45,10 @@ jobs:
+        TELEGRAM_USERNAME: "hhio618_bot"
+        TELEGRAM_TOKEN: "1008309528:AAGsWuc9UJ6Vu6dKnpw45MWO10kjpqLBjM8"
+        HCAPTCHA_SECRET: "0x8307B15a812B8B20033458dE663d89928E28D5a0"
+        HCAPTCHA_SITE_KEY: "bce377d8-1af3-464d-b9a7-16884a8cbc14"
```
Can you revert these to the fake ones? The real solution is to mock the tests, we don't want to keep real creds in this repo

hhio618 replied:
Ok, no problem :+1:

gaieges commented:
This wasn't quite noted in the original spec, but we don't need the concept of an "owner" as you have via the `TELEGRAM_OWNER_USERNAME` env var (and related code).  Everyone should be able to see the global stats, so lets get rid of that check/feature for now.

Also what we may do down the road is:

- Add the ability to break out what [users, channels] perform authentication via args to the stats command
- Add the ability to send this stats report to channels or users on some frequency.

That's more just an FYI, not in scope for this PR

gaieges commented on .github/workflows/pytest.yml:
```diff
@@ -3,12 +3,7 @@
-  pull_request:
-    # The branches below must be a subset of the branches above
-    branches: [master, staging]
+on: push
```
This part shouldnt be changing

hhio618 replied:
Ok, done

gaieges commented:
Deployed to staging environment and it looks good.. Unknown country however, know why that is? 

```
Current pool stats:
Humans: 100.00%(3)
Bots: 0.00%(0)
Percent of users, per country:
unknown: 100.00%
```

Why dont you fix that last test change mentioned above and we can close this out.

hhio618 commented:
> Deployed to staging environment and it looks good.. Unknown country however, know why that is?


If there is no known location information from Telegram, so we take those as unknown.

gaieges commented on .github/workflows/pytest.yml:
```diff
@@ -4,11 +4,6 @@
-    branches: '**'
-  pull_request:
-    # The branches below must be a subset of the branches above
-    branches: [master, staging]
```
Still appears that you're deleting this

hhio618 replied:
Ok, this one was fixed too. strange because I thought that I checked it out from upstream

gaieges replied:
Thanks!

gaieges commented:
We expect that test to fail, otherwise looking good!  Will merge and close out the bounty
